rinstall: refactoring and new environment variables

vorakl opened the pull request:
- Brake down the code into functions
- Add RINSTALL_DIFF_ARGS for modifying behaviour of diff(1) tool
- Add -a option for defining and alternative location of a file
- Saving a file in $SD doesn't depend on a current directory any more
- Fix: when diff(1) fails it is not treated as "files are different"
- Improve: remove curl(1) error messages where they are not informative

vorakl commented:
@eradman , this an initial version of the big change that brings a significant modification in the code structure although the code itself hasn't been changed. I just rearranged it. That's why I recommend to look at the full listing of a new file rather than trying to compare in two sides diff's output.
`main()` function becomes a workflow, a high level list of steps. All the details are moved into functions.

Tell me if something needs improvements.

I also added those additional features (RINSTALL_DIFF_ARGS and ALTERNATIVE_SOURCE) as it turned out that after the refactoring, it's just a little addition to the code.

eradman reviewed (changes requested):
Nicely organized code!

My inclination is to make `ALTERNATIVE_SOURCE` mirror `INSTALL_URL`.

In this example check for `wordpress-6.4.2.tar.gz` would be pulled from `_sources` using the built-in web server first, then wordpress.org

```
export ALTERNATIVE_SOURCE="https://wordpress.org/"
$SD/rinstall wordpress-6.4.2.tar.gz
```

With this behavior then it would probably make more sense to call this environment variable `ALTERNATIVE_INSTALL_URL`.

vorakl commented on tests/test_rinstall.rb:
```diff
@@ -80,9 +81,13 @@ def eq(result, expected)
   File.open(src, 'w') { |f| f.write('123') }
   cmd = "INSTALL_URL=#{@install_url} #{Dir.pwd}/../rinstall -m 644 x/y/#{fn} #{dst}"
   out, err, status = Open3.capture3(cmd, chdir: @systmp)
+puts "out: [#{out}]"
```
yep, I've noticed that ;) and fixed just 1 minute before your message. So, it's fixed already.

vorakl commented:
My intention was a bit different: give an option to download from absolutely uncontrolled (3rd party) source, where files organization is not only different, but also may change anytime.
I see what you mean, but I think, what you suggest is a great case for one more variable - BACKUP_SOURCE
Let me explain on an example.

The latest merged change brought a significant modification in how Staging Directory is used.
For instance,  `$SD/rinstall etc/ssh/sshd_conf` goes through the following logic:

 1. Makes sure `$SD/etc/ssh/` exists. If it exists, then rset copied `etc/ssh/sshd_conf` to $SD with `tar` already
 2. Checks if `$(pwd)/etc/ssh/sshd_config` exists. Right now, rset keeps $PWD and $SD the same so, it effectively checks for the existence of `$SD/etc/ssh/sshd_conf`. If  source is defined with an absolute path, it checks on the local file system on a remote host. But if it's a relative path, then it checks in $SD, which also happens to be a current directory. If it doesn't exist, then this file wasn't copied to $SD with `tar`. The example uses a relative path so, it doesn't check on a file system.  Next step is to try to download it from `_sources` (using $INSTALL_URL)
 3. When rset tries to download from `_source/` , it uses HTTP location `$INSTALL_URL/etc/ssh/sshd_conf`. And this is important, because it represents a transparency between a file structure provided with `tar` and a file structure maintained in `_source` and used with HTTP request. Another important note, both file structures are under our control and reside under the same  directory on a control host (where rset is executed). So, `etc/ssh/` part of the path is the same in both checks.

If `ALTERNATIVE_SOURCE` mirrors `INSTALL_URL`, then it will also be a resource under our control, because sshd should reside  in `$ALTERNATIVE_SOURCE/etc/ssh/` location.
But how would we get a file from a 3rd party URL if it is  `somehost.org/pub/releases/1.2.3/patch-19/configuration/defaults/sshd_config` ? No way. Or, how can I define a location for, let's say, a huge archive with installation from a typical "corporate" location with a dozen of sub-folders, if I didn't cached it in `_source`? No way.
And, another argument to support this workflow, take a look at the example I added into man. It shows how to download any file from an external resource in a system agnostic way, meaning, you don't need to think what the OS is and what the tool (wget/curl/fetch/...) needs to be used, you just run this

`ALTERNATIVE_SOURCE="https://example.org/dl/sometool-x86_64-$(uname).bin" $SD/rinstall _ /usr/local/bin/sometool`

This behavior is a key for a sharable deployments, where I can avoid making a giant archive by excluding binaries in a tgz  and just sharing the code and it'll still be working because there is an ALTERNATIVE location to fetch them all.

My point is, the thing you suggest covers a bit different case. I'd leave ALTERNATIVE_SOURCE with its unique use case and add BACKUP_SOURCE to try it before ALTERNATIVE_SOURCE which is another resource under our control.

eradman commented:
In that case, couldn't this be an argument instead of an environment variable?

```
$SD/rinstall sometool.bin -a https://example.org/dl/sometool-x86_64-$(uname).bin /usr/local/bin/sometool
```

vorakl commented:
Yes, absolutely because it is always a full URL to a file.
It'd work for me, I don't see any issues.

So, rewrite the code to replace ALTERNATIVE_SOURCE with `-a` cli option?
What about a mirror for INSTALL_URL?  Add it either? May be even MIRROR_INSTALL_URL?

vorakl commented:
Yep, I think, command line option is definitely better than an environment variable, because that "alternative" source cannot be reused for other downloads and it doesn't make sense to export it or defined in `environment=`. It is something that has meaning only per one rinstall command.

eradman commented:
Yes, go ahead and replace `ALTERNATIVE_SOURCE` with `-a` CLI option.

We can include `MIRROR_INSTALL_URL` feature in a future change.

vorakl commented:
I've rewrote with '-a' option.
I've also caught a few other edge cases and issues.
The biggest improvement is that now `rinstall` doesn't depend on a CWD at all. Before this patch, it was silently assumed that `rset` sets CWD to $SD for rinstall. But, this is s a very weak assumption and if CWD and SD were not the same, the whole logic of the script failed. SO now, CWD doesn't play any role, it can be anything and `rinstall` behaves correctly.

I've also stressed out what was partially assumed and defined in the documentation:  if the source is defined with an absolute path then it is located on a remote's file system. But if it's a relative path, then it either already resides in SD (from a previous `rinstall` run or brought in by `tar`) or is located or a remote HTTP resource. If non of these actions get the source, `rinstall` fails clearly

There was also an interesting bug in how diff(1) was used. In case of the diffs' fail (exit status 2) it was mistakenly accepted as "files are different"... even where there were no files ;)

Ans cp(1) now correctly locates the source depending on its definition (local or remote over HTTP).

I've updated a PR description and man pages as well.

eradman commented:
I like the simplicity and self-documenting nature of this this:

```
$SD/rinstall -a https://wordpress.org/wordpress-6.4.2.tar.gz wordpress.tar.gz
```

This is the output (on OpenBSD) if not found in `_sources`:

```
ftp: Error retrieving http://127.0.0.1:6000/wordpress.tar.gz: 404 Not Found
rinstall: fetched /tmp/rset_staging_10931/wordpress.tar.gz
```

I think it would be helpful to print a message before the download indicating that rinstall is trying thalternate URL:

```
ftp: Error retrieving http://127.0.0.1:6000/wordpress.tar.gz: 404 Not Found
rinstall: using alternate source: https://wordpress.org/wordpress-6.4.2.tar.gz
rinstall: fetched /tmp/rset_staging_10931/wordpress.tar.gz
```

@vorakl, what do you think?

vorakl commented:
> This is the output (on OpenBSD) if not found in _sources:

I faced exactly this behavior with curl and removed `-S` that actually solved it. I didn't want to hide error messages at all. Luckily, curl simply avoids extra error printing but still uses the correct exit codes. By some reason, I didn't think about other tools, and didn't know that `ftp` is also affected... ;|

> I think it would be helpful to print a message before the download indicating that `rinstall` is trying the alternate URL:

Sure, such a message will make it clear for everyone, no matter what a tool  is used to download it.
I'll submit a change in shortly.

vorakl commented:
Done.

```
$ SD="/tmp/rset" sh ./rinstall.sh -a http://localhost:8080/etc/passwd etc/passwd 
rinstall: using alternate source: http://localhost:8080/etc/passwd
rinstall: fetched /tmp/rset/etc/passwd
```

vorakl commented:
> I like the simplicity and self-documenting nature of this this:

I agree! It surprisingly gets into place like it's always been there :)

eradman commented on rinstall.1:
```diff
@@ -50,10 +51,20 @@ is omitted, the
+.Ev INSTALL_URL
+or
+.Fl a
+option that defines an aternative file locataion.
```
Suggested rewording:

> fetch it from the HTTP resource defined by `INSTALL_URL` or the `-a` option.

eradman commented on rinstall.1:
```diff
@@ -68,9 +79,18 @@ This argument is passed to
-.Xr chown 8 .
+.Xr chown 1 .
+.It Fl a
+URL to an alternative file location
```
End with a `.`

eradman commented on rinstall.sh:
```diff
@@ -2,143 +2,244 @@
+	check_absolute_path "$src_path" && {
+		>&2 echo "rinstall: absolute path is not allowed for HTTP fetch: $source"
+		exit 1
+	} || :
```
Is this nessesary?

`set -e` /  `trap ERR` is not used so I think it's okay for individual statements to exit non-zero

vorakl replied:
I agree that it looks a bit cryptic. I rewrote quickly all similar conditions with if-statement.

eradman reviewed (changes requested):
Only small items. Everything seems to be in good working order

Tested on Alpine Linux, Ubuntu 22, FreeBSD 12, OpenBSD 7.4

eradman commented:
Looks good!
